File: T1.SqlSharp/T1.SqlSharp/Expressions/SqlJoinTableCondition.cs
using T1.Standard.IO;

namespace T1.SqlSharp.Expressions;

public class SqlJoinTableCondition : ISqlExpression 
{
    public SqlType SqlType { get; } = SqlType.JoinCondition; 
    public TextSpan Span { get; set; } = new();
    public void Accept(SqlVisitor visitor)
    {
        visitor.Visit_JoinTableCondition(this); 
    }

    public JoinType JoinType { get; set; } = JoinType.Inner;
    public required ITableSource JoinedTable { get; set; }
    public ISqlExpression? OnCondition { get; set; }
    public string ToSql()
    {
        var sql = new IndentStringBuilder();
        sql.Write(GetJoinKeyword());
        sql.Write(" ");
        sql.Write(JoinedTable.ToSql());
        if (OnCondition != null)
        {
            sql.Write(" ON ");
            sql.Write(OnCondition.ToSql());
        }
        return sql.ToString();
    }

    private string GetJoinKeyword()
    {
        return JoinType switch
        {
            JoinType.Cross => "CROSS JOIN",
            JoinType.CrossApply => "CROSS APPLY",
            JoinType.OuterApply => "OUTER APPLY",
            _ => $"{JoinType.ToString().ToUpper()} JOIN"
        };
    }
}

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlJoinTableCondition.ts
import { SqlType } from './SqlType';
import { TextSpan } from '../StringParser';
import { ISqlExpression, SqlVisitor } from './ISqlExpression';
import { JoinType } from './JoinType';
import { ITableSource } from './ITableSource';

export class SqlJoinTableCondition implements ISqlExpression {
    SqlType: SqlType = SqlType.JoinCondition;
    Span: TextSpan = new TextSpan();
    JoinType: JoinType = JoinType.Inner;
    JoinedTable!: ITableSource;
    OnCondition!: ISqlExpression;

    Accept(visitor: SqlVisitor): void {
        visitor.Visit_JoinTableCondition(this);
    }

    ToSql(): string {
        let sql = this.JoinType.toString().toUpperCase() + ' JOIN ';
        sql += this.JoinedTable.ToSql();
        sql += ' ON ' + this.OnCondition.ToSql();
        return sql;
    }
} 